flink-ml-benchmark/src/main/java/org/apache/flink/ml/benchmark/BenchmarkUtils.java

```diff
@@ -0,0 +1,233 @@
+     * environment.
+     *
+     * @return Results of the executed benchmark.
+     */
+    public static BenchmarkResult runBenchmark(
+            String name, StreamTableEnvironment tEnv, Map<String, ?> benchmarkParamsMap)
+            throws Exception {
+        Stage<?> stage =
+                (Stage<?>) instantiateWithParams((Map<String, ?>) benchmarkParamsMap.get("stage"));
+
+        BenchmarkResult result;
+        if (benchmarkParamsMap.size() == 2) {
+            DataGenerator<?> inputsGenerator =
+                    (DataGenerator<?>)
+                            instantiateWithParams(
+                                    (Map<String, ?>) benchmarkParamsMap.get("inputs"));
+            result = runBenchmark(name, tEnv, stage, inputsGenerator);
+        } else if (benchmarkParamsMap.size() == 3 && stage instanceof Model) {
+            DataGenerator<?> inputsGenerator =
+                    (DataGenerator<?>)
+                            instantiateWithParams(
+                                    (Map<String, ?>) benchmarkParamsMap.get("inputs"));
+            DataGenerator<?> modelDataGenerator =
+                    (DataGenerator<?>)
+                            instantiateWithParams(
+                                    (Map<String, ?>) benchmarkParamsMap.get("modelData"));
+            result =
+                    runBenchmark(name, tEnv, (Model<?>) stage, modelDataGenerator, inputsGenerator);
+        } else {
+            throw new IllegalArgumentException(
+                    "Unsupported json map with keys " + benchmarkParamsMap.keySet());
+        }
+
+        return result;
+    }
+
+    /**
+     * Executes a benchmark from a stage and an inputsGenerator in the provided environment.
+     *
+     * @return Results of the executed benchmark.
+     */
+    public static BenchmarkResult runBenchmark(
+            String name,
+            StreamTableEnvironment tEnv,
+            Stage<?> stage,
+            DataGenerator<?> inputsGenerator)
+            throws Exception {
+        Table[] inputTables = inputsGenerator.getData(tEnv);
+
+        Table[] outputTables;
+        if (stage instanceof Estimator) {
+            outputTables = ((Estimator<?, ?>) stage).fit(inputTables).getModelData();
+        } else if (stage instanceof AlgoOperator) {
+            outputTables = ((AlgoOperator<?>) stage).transform(inputTables);
+        } else {
+            throw new IllegalArgumentException("Unsupported Stage class " + stage.getClass());
+        }
+
+        for (Table table : outputTables) {
+            tEnv.toDataStream(table).addSink(new DiscardingSink<>());
+        }
+
+        StreamExecutionEnvironment env = ((StreamTableEnvironmentImpl) tEnv).execEnv();
+        JobExecutionResult executionResult = env.execute();
+
+        BenchmarkResult result = new BenchmarkResult();
+        result.name = name;
+        result.executionTimeMillis = (double) executionResult.getNetRuntime(TimeUnit.MILLISECONDS);
```

Comment:
We need a general solution to calculate the throughput.

How about this:
- Expose API in the DataGenerator to get the number of input values used in the fit/transform.
- Get the number of output values from outputTables.
- Expose the inputValueNum, outputNum, inputThoughput and outputThroughput in the BenchmarkResult.

Reply:
I can make `DataGenerator implements GeneratorParams`, which means each generator must have a `getNumData()` method. That should be a simpler solution.

Reply:
The alternative approach sounds good.